src/main/java/org/sonarlint/intellij/server/SonarLintHttpServer.kt

```diff
@@ -0,0 +1,182 @@
+import io.netty.handler.logging.LogLevel
+import io.netty.handler.logging.LoggingHandler
+import io.netty.util.CharsetUtil
+import org.sonarlint.intellij.util.SonarLintUtils
+import java.net.BindException
+import kotlin.concurrent.thread
+
+const val STARTING_PORT = 63000
+const val INVALID_REQUEST = "Invalid request."
+const val UNKNOWN_INTELLIJ_FLAVOR = "Unknown IntelliJ flavor."
+const val PORT_RANGE = 3
+const val ENVIRONMENT_ENDPOINT = "/sonarlint/environment"
+
+
+
+object SonarLintHttpServer {
+
+    fun start() {
+        tryToStart(0)
+    }
+
+    fun tryToStart(numberOfAttempts: Int) {
+        try {
+            actualStart(STARTING_PORT + numberOfAttempts)
+        } catch (e: BindException) {
+            if (numberOfAttempts < PORT_RANGE) {
+                tryToStart(numberOfAttempts + 1)
+            } else {
+                throw RuntimeException("Couldn't start SonarLint server in port range: $STARTING_PORT - ${STARTING_PORT + PORT_RANGE}")
+            }
+        }
+    }
+
+    fun actualStart(port: Int) {
+        // don't block UI thread
+        thread {
+            val bossGroup: EventLoopGroup = NioEventLoopGroup(1)
+            val workerGroup: EventLoopGroup = NioEventLoopGroup()
+            try {
+                val b = ServerBootstrap()
+                b.group(bossGroup, workerGroup)
+                        .channel(NioServerSocketChannel::class.java)
+                        .handler(LoggingHandler(LogLevel.INFO))
+                        .childHandler(HttpSnoopServerInitializer())
+                val ch = b.bind(port).sync().channel()
+                System.err.println("Open your web browser and navigate to http://localhost:$port/")
+                ch.closeFuture().sync()
+                System.err.println("After close future")
+            } catch (e: Exception) {
+                System.err.println("Actual start error: ${e.message}")
+                throw e
+            } finally {
+                System.err.println("Finally after server start.")
+                bossGroup.shutdownGracefully()
+                workerGroup.shutdownGracefully()
+            }
+        }
+    }
+}
+
+class HttpSnoopServerInitializer : ChannelInitializer<SocketChannel?>() {
+    override fun initChannel(ch: SocketChannel?) {
+        ch ?: return
+        val p = ch.pipeline()
+        p.addLast(HttpRequestDecoder())
+        p.addLast(HttpResponseEncoder())
+        p.addLast(HttpSnoopServerHandler())
+    }
+}
+
+class HttpSnoopServerHandler : SimpleChannelInboundHandler<Any?>() {
+    private var request: HttpRequest? = null
+
+    private val buf = StringBuilder()
+    override fun channelReadComplete(ctx: ChannelHandlerContext) {
+        ctx.flush()
+    }
+
+
+    private fun processRequest(request: HttpRequest): String {
+        if (request.uri() == ENVIRONMENT_ENDPOINT && request.method() == HttpMethod.GET) {
+            return SonarLintUtils.getIdeVersionForTelemetry() ?: UNKNOWN_INTELLIJ_FLAVOR
```

Comment:
The handler for the request and the response that we return are hidden in the middle of this big server class, mixed with details about the server setup. Could we find a way to extract this somewhere else ?